backend/http/http.go

```diff
@@ -0,0 +1,345 @@
+package http
+
+import (
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v6/plumbing/storer"
+	"github.com/go-git/go-git/v6/plumbing/transport"
+	"github.com/go-git/go-git/v6/storage"
+	"github.com/go-git/go-git/v6/utils/ioutil"
+)
+
+type contextKey string
+
+type service struct {
+	pattern  *regexp.Regexp
+	method  string
+	handler http.HandlerFunc
+	svc     transport.Service
+}
+
+var services = []service{
+	{regexp.MustCompile("(.*?)/HEAD$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/info/refs$"), http.MethodGet, getInfoRefs, ""},
+	{regexp.MustCompile("(.*?)/objects/info/alternates$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/objects/info/http-alternates$"), http.MethodGet, getTextFile, ""},
+	{regexp.MustCompile("(.*?)/objects/info/packs$"), http.MethodGet, getInfoPacks, ""},
+	{regexp.MustCompile("(.*?)/objects/[0-9a-f]{2}/[0-9a-f]{38}$"), http.MethodGet, getLooseObject, ""},
+	{regexp.MustCompile("(.*?)/objects/[0-9a-f]{2}/[0-9a-f]{62}$"), http.MethodGet, getLooseObject, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{40}\\.pack$"), http.MethodGet, getPackFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{64}\\.pack$"), http.MethodGet, getPackFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{40}\\.idx$"), http.MethodGet, getIdxFile, ""},
+	{regexp.MustCompile("(.*?)/objects/pack/pack-[0-9a-f]{64}\\.idx$"), http.MethodGet, getIdxFile, ""},
+
+	// TODO: Support git-upload-archive
+	// {regexp.MustCompile("(.*?)/git-upload-archive$"), http.MethodPost, serviceRpc, transport.UploadArchiveService},
+	{regexp.MustCompile("(.*?)/git-upload-pack$"), http.MethodPost, serviceRpc, transport.UploadPackService},
+	{regexp.MustCompile("(.*?)/git-receive-pack$"), http.MethodPost, serviceRpc, transport.ReceivePackService},
+}
+
+// DefaultLoader is the default loader used to load repositories from storage.
+// It will use the current working directory as the base path for the
+// repositories.
+var DefaultLoader = transport.NewFilesystemLoader(osfs.New("."), false)
+
+// Handler is a Git HTTP handler that serves git repositories over HTTP.
+//
+// It supports serving repositories using both the Smart-HTTP and the Dumb-HTTP
+// protocols. When the Dumb-HTTP protocol is used, the repository store must
+// implement the [storer.FilesystemStorer] interface. Keep in mind that
+// repositories that wish to be server using the Dumb-HTTP protocol must update
+// the server info files. This can be done by using
+// [transport.UpdateServerInfo] before serving the repository.
+type Handler struct {
```

Comment:
From an API/UX perspective, the public fields can be changed for a created `Handler`. We are better off making all the fields internal, and have them set via the[ variadic options pattern](https://github.com/go-git/go-git/blob/93ed29d8f65c1f879d438a5370c212ffedfe5030/plumbing/format/packfile/parser_test.go#L39-L40) based on a `NewHandler(opts ...)` instead.